### PR TITLE
Avoid presence of our internal types in generated code

### LINKED ISSUE #295: Bad types choosing
**Description**

There are random crashes from Resolver when users class contains Generics. It happens because TypeResolver::findAnyConcreteInheritorIncludingOrDefaultUnsafe can return non user's class type, for example, any class from soot.

**To Reproduce**

It's impossible to reproduce always because there are some hacks to reduce number of situations like that. I detect it while testing on IDEA plugin class StandardMutableValueGraphTest from guava.

**Expected behavior**

TypeResolver returns only classes that are available to user.

**Actual behavior**

TypeResolver::findAnyConcreteInheritorIncludingOrDefaultUnsafe can return non user's class type.

**Visual proofs (screenshots, logs, images)**

~~~
java.lang.ClassNotFoundException: soot.IntType
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:471)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:589)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at org.utbot.engine.Resolver.constructModel(Resolver.kt:513)
	at org.utbot.engine.Resolver.resolveObject(Resolver.kt:465)
	at org.utbot.engine.Resolver.resolveReferenceValue(Resolver.kt:424)
	at org.utbot.engine.Resolver.resolveModel(Resolver.kt:395)
	at org.utbot.engine.Resolver.collectMockBehaviour(Resolver.kt:601)
	at org.utbot.engine.Resolver.resolveMock(Resolver.kt:341)
	at org.utbot.engine.Resolver.resolveReferenceValue(Resolver.kt:416)
	at org.utbot.engine.Resolver.resolveModel(Resolver.kt:395)
	at org.utbot.engine.Resolver.internalResolveModel(Resolver.kt:233)
	at org.utbot.engine.Resolver.resolveModels$utbot_framework(Resolver.kt:204)
	at org.utbot.engine.UtBotSymbolicEngine$traverseImpl$1.invokeSuspend(UtBotSymbolicEngine.kt:471)
~~~

### PR DESCRIPTION
**Pull request by CaelmBleidd**
# Description

This request remove soot types and our internal types from soot classes at all, and our classes used for mocks, overrides, etc., from the construction of type storages. At the same time, it changes API for `TypeStorage` class to avoid misunderstanding of what types will it contains after its creation via different constructors. 

Fixes #295 

## Type of Change
- Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

## Automated Testing

There was no specific tests added. This request should not break already existed as a proof that we didn't remove anything that was required by our overridden classes. 

## Manual Scenario 

There was no manual testing. 

# Checklist:

- [x] The change followed the style guidelines of the UTBot project
- [x] Self-review of the code is passed
- [x] The change contains enough commentaries, particularly in hard-to-understand areas
- [x] New documentation is provided or existed one is altered
- [x] No new warnings
- [x] All tests pass locally with my changes

### CODE REVIEW

File: utbot-framework/src/main/kotlin/org/utbot/engine/DataClasses.kt [hunk 1]
```diff
@@ -121,6 +124,29 @@ data class TypeStorage(val leastCommonType: Type, val possibleConcreteTypes: Set
...
+         *
+         * In regular cases you should use [TypeResolver.constructTypeStorage] method instead.
+         */
+        fun constructTypeStorageUnsafe(
+            leastCommonType: Type,
+            possibleConcreteTypes: Set<Type>
+        ): TypeStorage = TypeStorage(leastCommonType, possibleConcreteTypes)
+
+        /**
+         * Constructs a type storage with some type. In this case [possibleConcreteTypes] might contains
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/engine/DataClasses.kt:143]: `might contain`

File: utbot-framework/src/main/kotlin/org/utbot/engine/DataClasses.kt [hunk 2]
```diff
@@ -121,6 +124,29 @@ data class TypeStorage(val leastCommonType: Type, val possibleConcreteTypes: Set
...
+         * abstract class or interface. Usually it means such typeStorage represents wrapper object type.
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/engine/DataClasses.kt:144]: `an abstract ...`

File: utbot-framework/src/main/kotlin/org/utbot/engine/Memory.kt [hunk 3]
```diff
@@ -365,8 +365,12 @@ data class Memory( // TODO: split purely symbolic memory and information about s
  */
 class TypeRegistry {
     init {
+        // TODO mistake
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/engine/Memory.kt:368]: Perhaps we should create a task to fix it someday

File: utbot-framework/src/main/kotlin/org/utbot/engine/Memory.kt [hunk 4]
```diff
@@ -612,9 +616,11 @@ class TypeRegistry {
     fun createClassRef(baseType: Type, numDimensions: Int = 0): MethodResult {
         val addr = classRefBiMap.getOrPut(baseType) { nextClassRefAddr() }
 
-        val objectValue = ObjectValue(TypeStorage(CLASS_REF_TYPE), addr)
+        val objectTypeStorage = TypeStorage.constructTypeStorageWithSingleType(CLASS_REF_TYPE)
+        val objectValue = ObjectValue(objectTypeStorage, addr)
 
-        val typeConstraint = typeConstraint(addr, TypeStorage(CLASS_REF_TYPE)).all()
+        val classRefTypeStorage = TypeStorage.constructTypeStorageWithSingleType(CLASS_REF_TYPE)
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/engine/Memory.kt:622]: May we use the type storage above?

File: utbot-framework/src/main/kotlin/org/utbot/framework/util/SootUtils.kt [hunk 5]
```diff
@@ -89,9 +89,31 @@ private fun initSoot(buildDirs: List<Path>, classpath: String?, jdkInfo: JdkInfo
...
+        val isUtBotPackage = it.packageName.startsWith(UTBOT_PACKAGE_PREFIX)
+
+        // remove our own classes from the soot scene
+        if (isUtBotPackage) {
+            val isOverriddenPackage = it.packageName.startsWith(UTBOT_OVERRIDDEN_PACKAGE_PREFIX)
+            val isExamplesPackage = it.packageName.startsWith(UTBOT_EXAMPLES_PACKAGE_PREFIX)
+            val isApiPackage = it.packageName.startsWith(UTBOT_API_PACKAGE_PREFIX)
+
+            // remove if it is not a part of the examples (CUT), not a part of our API and not an override
+            if (!isOverriddenPackage && !isExamplesPackage && !isApiPackage) {
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/util/SootUtils.kt:102]: It looks like regular users do not need classes from utbot samples. Can we detect such situations somehow?

File: utbot-framework/src/main/kotlin/org/utbot/framework/util/SootUtils.kt [hunk 6]
```diff
@@ -89,9 +89,31 @@ private fun initSoot(buildDirs: List<Path>, classpath: String?, jdkInfo: JdkInfo
...
+                Scene.v().removeClass(it)
+                return@forEach
+            }
+        }
+
+        // remove soot's classes from the scene, because we don't wont to analyze them
+        if (it.packageName.startsWith(SOOT_PACKAGE_PREFIX)) {
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/util/SootUtils.kt:109]: Just curious - odes it mean that we will not be able to analyze Soot itself?

**Review by Damtev — CHANGES_REQUESTED** [review]

File: utbot-framework/src/main/kotlin/org/utbot/engine/Memory.kt (hunk 3 above)
Comment (CaelmBleidd) [reply utbot-framework/src/main/kotlin/org/utbot/engine/Memory.kt:368]: We have it https://github.com/UnitTestBot/UTBotJava/issues/1007, I forgot to remove this comment

File: utbot-framework/src/main/kotlin/org/utbot/framework/util/SootUtils.kt (hunk 5 above)
Comment (CaelmBleidd) [reply utbot-framework/src/main/kotlin/org/utbot/framework/util/SootUtils.kt:102]: I think no, but I hope that they will be out of classpath during the analysis on real code (but I'm not sure)

File: utbot-framework/src/main/kotlin/org/utbot/framework/util/SootUtils.kt (hunk 6 above)
Comment (CaelmBleidd) [reply utbot-framework/src/main/kotlin/org/utbot/framework/util/SootUtils.kt:109]: Yes, I'll make an option for it

File: utbot-framework-api/src/main/kotlin/org/utbot/framework/UtSettings.kt [hunk 7]
```diff
@@ -382,6 +382,18 @@ object UtSettings : AbstractSettings(
...
+
+    /**
+     * If this options set in true, all soot classes will be removed from a Soot Scene,
+     * therefore, you will be unable to test soot classes.
+     */
+    var removeSootClassesFromHierarchy by getBooleanProperty(true)
+
+    /**
+     * If this options set in true, all soot classes will be removed from a Soot Scene,
+     * therefore, you will be unable to test soot classes.
```
Comment (Damtev) [inline utbot-framework-api/src/main/kotlin/org/utbot/framework/UtSettings.kt:394]: Not soot but utbot

**Review by Damtev — CHANGES_REQUESTED** [review]

**Review by Damtev — APPROVED** [review]

**Pull request merged**